### DISCUSSION TITLE
Unsafe.SkipInit and out parameter

### DISCUSSION THREAD

**Discussion by sakno** [dotnet/runtime#57490] [Q&A] [answered]
Many parsing API looks like following:
```csharp
public static bool TryParse(string input, out SomeType result)
{
}
```

When method returns **false**, is it allowed to use `Unsafe.SkipInit(out result)` instead of `result = default`? Most callers ignore the result if it returns **false** so leaving it uninitialized can save some instructions cause by **initobj** IL instruction.

Moreover, if `SomeType` is a reference type so it will be initialized as **null** by the caller.

**Comment by EgorBo** [comment, marked as answer]
In my view, cases where TryParse fails are (expected to be) rare so it's not that important to optimize. And returning junk in `out` doesn't look safe. E.g.
```csharp
TryParse(input, out MyStruct ms);
if (ms.Id != 0) // Id will be some random value in case of failed TryParse
{
...
```

**Reply by stephentoub** [reply]
Additionally, there's also a good deal of code that expects a default value to come out when a Try method returns false.  Having that default value be garbage would likely lead to a lot of subtle bugs.  It would also propagate the "unsafe"ness out to callers without there being any indication that's happening, in that you can then end up with instances that violate whatever constraints their value type constructors try to enforce.

**Reply by IS4Code** [reply]
More important than what the user might expect from an unused `out` parameter is what the documentation tells about the method. Most `TryParse` methods specify that the argument is set to `default`. This would be a breaking change then.
